### PR TITLE
Add support for Pydantic v1 and v2.

### PR DESCRIPTION
**Pull request by ddelange**
*Issue #, if available:*

*Description of changes:*

No breaking or functional changes, continues to use the v1 codebase that is shipped with pydantic v2.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

**Please tag this pr with at least one of these labels to make our release process faster:** BREAKING, new feature, bug fix, other change, dev setup

### CODE REVIEW

**Comment by lostella** [comment]
@ddelange thanks for this! I fixed a couple of details that were missing. Kindly asking @jaheba to chime in to review this

File: src/gluonts/meta/_version.py
Comment (jaheba) [inline src/gluonts/meta/_version.py:1]: Can we have this in a separate PR?

**Comment by jaheba** [comment]
Thanks for the PR!

Maybe it is easier to define a new `pydantic` module from which we import everywhere else. A bit like there is `gluonts.json` as a facade for different json implementations.

That way we might also be able to align on the type and thus make mypy happy.

File: src/gluonts/meta/_version.py
Comment (ddelange) [reply src/gluonts/meta/_version.py:1]: https://github.com/awslabs/gluonts/pull/3028

**Comment by ddelange** [comment]
Hi @jaheba 👋

I've moved the VCS commit out and added a commit with your suggestion. Please have a look:)

**Comment by ddelange** [comment]
There's one mypy error left on my machine that seems unrelated to this PR (but may be due to the versions installed on my system):
```
src/gluonts/zebras/_time_frame.py:449: note: By default the bodies of untyped functions are not checked, consider using --check-untyped-defs  [annotation-unchecked]
src/gluonts/zebras/_time_frame.py:453: note: By default the bodies of untyped functions are not checked, consider using --check-untyped-defs  [annotation-unchecked]
src/gluonts/mx/model/gp_forecaster/gaussian_process.py:395: error: Argument 1 to "axis" has incompatible type "list[Any]"; expected "tuple[float, float, float, float] | bool | str | None"  [arg-type]
Found 1 error in 1 file (checked 249 source files)
```

edit: green on CI 🤔

**Comment by jaheba** [comment]
Looks good to me!

The only thing I would add is an `__all__` in `pydantic.py`, which contains all exposed names. Then we also won't need the `# noqa` anymore.

**Comment by jaheba** [comment]
@lostella Not sure we should backport this, since this could potentially be breaking?

**Comment by lostella** [comment]
@jaheba 0.14 is still not out (only rc) so we can still allow some breaking maybe? how breaking is this?

**Comment by ddelange** [comment]
@lostella no breaking changes here, `pydantic.v1` is a 1-1 replacement for v1 `pydantic` :)

**Review by jaheba — DISMISSED** [review]

**Comment by lostella** [comment]
merge pending additional changes from @jaheba to have the docs build ignore problematic modules/packages:
- `gluonts.pydantic`
- `gluonts.nursery`

**Comment by jaheba** [comment]
Let's merge #3030 first, then add `pydantic` in the ignore section of the build as well.

**Comment by lostella** [comment]
#3030 and #3031 are merged, @ddelange could you add similar changes to exclude the pydantic module from the docs build? That should get rid of the failures, and then we can merge 🚀

**Comment by ddelange** [comment]
Like that?

**Comment by lostella** [comment]
Amazing, thanks a lot for the PR @ddelange!

**Review by lostella — APPROVED** [review]

**Pull request merged**